runtime/engine/src/main/java/io/aklivity/zilla/runtime/engine/internal/registry/HttpWatcherTask.java

```diff
@@ -0,0 +1,260 @@
+    }
+
+    @Override
+    public Future<Void> submit()
+    {
+        return executor.submit(this);
+    }
+
+    @Override
+    public Void call() throws InterruptedException
+    {
+        while (true)
+        {
+            URI configURI = configWatcherQueue.take();
+            if (configURI == CLOSE_REQUESTED)
+            {
+                break;
+            }
+            String etag = etags.getOrDefault(configURI, "");
+            sendAsync(configURI, etag);
+        }
+        return null;
+    }
+
+    @Override
+    public CompletableFuture<NamespaceConfig> watch(
+        URL configURL)
+    {
+        URI configURI = toURI(configURL);
+        NamespaceConfig config = sendSync(configURI, INTIAL_ETAG);
+        if (config == null)
+        {
+            return CompletableFuture.failedFuture(new Exception("Parsing of the initial configuration failed."));
+        }
+        return CompletableFuture.completedFuture(config);
+    }
+
+    @Override
+    public String readURL(
+        URL configURL)
+    {
+        String output = "";
+        HttpClient client = HttpClient.newBuilder()
+            .version(HTTP_2)
+            .followRedirects(NORMAL)
+            .build();
+        HttpRequest request = HttpRequest.newBuilder()
+            .GET()
+            .uri(toURI(configURL))
+            .build();
+
+        try
+        {
+            HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
+            output = response.body();
+        }
+        catch (Exception ex)
+        {
+            rethrowUnchecked(ex);
+        }
+        return output;
+    }
+
+    @Override
+    public void close()
+    {
+        futures.values().forEach(future -> future.cancel(true));
+        configWatcherQueue.add(CLOSE_REQUESTED);
+    }
+
+    private NamespaceConfig sendSync(
+        URI configURI,
+        String etag)
+    {
+        HttpClient client = HttpClient.newBuilder()
+            .version(HTTP_2)
+            .followRedirects(NORMAL)
+            .build();
+        HttpRequest request = HttpRequest.newBuilder()
+            .GET()
+            .headers("If-None-Match", etag, "Prefer", "wait=86400")
+            .uri(configURI)
+            .build();
+        HttpResponse<String> response;
+        try
+        {
+            response = client.send(request, HttpResponse.BodyHandlers.ofString());
+        }
+        catch (Exception ex)
+        {
+            handleException(ex, configURI);
+            return null;
+        }
+        return handleConfigChange(response);
+    }
+
+    private void sendAsync(
+        URI configURI,
+        String etag)
+    {
+        HttpClient client = HttpClient.newBuilder()
+            .version(HTTP_2)
+            .followRedirects(NORMAL)
+            .build();
+        HttpRequest request = HttpRequest.newBuilder()
+            .GET()
+            .headers("If-None-Match", etag, "Prefer", "wait=86400")
```

Comment:
Do we actually send `if-none-match` with value `""` or is that header omitted from the request when the etag is empty string?

Reply:
As far as I understood, the etag that we get back from the zilla config server should never be empty. Even if the file is empty, we'll get the hash of the empty byte[]. So no, we should never send it with empty string. At the beginning we're sending with "INIT" constant.

Reply:
`etag` is not a required HTTP response field, so you cannot assume it is always present.

Reply:
At the beginning, you should not be sending any value for `if-match`, rather than `INIT`.